src/freenet/client/filter/PNGFilter.java

```diff
@@ -237,20 +247,56 @@ public void readFilter(InputStream input, OutputStream output, String charset, M
 				if (!skip && "PLTE".equalsIgnoreCase(chunkTypeString)) {
 					if (hasSeenIDAT)
 						throwError("PLTE must be before IDAT", "PLTE must be before IDAT");
+					if (hasSeenMDCV && !hasSeenCICP) {
+						throwError("mDCV requires cICP", "mDCV requires cICP");
+					}
 					validChunkType = true;
+					hasSeenPLTE = true;
 				}
 
 				if (!skip && "IDAT".equalsIgnoreCase(chunkTypeString)) {
 					if (hasSeenIDAT && !"IDAT".equalsIgnoreCase(lastChunkType))
 						throwError("Multiple IDAT chunks must be consecutive!",
 						        "Multiple IDAT chunks must be consecutive!");
+					if (hasSeenMDCV && !hasSeenCICP) {
+						throwError("mDCV requires cICP", "mDCV requires cICP");
+					}
 					hasSeenIDAT = true;
 					validChunkType = true;
 				}
 
+				if (!skip && "cICP".equals(chunkTypeString)) {
+					if (length != 4)
+						throwError("cICP chunks invalid!",
+								"cICP chunks must be 4 bytes long!");
+					if (chunkData[2] != 0)
+						throwError("cICP chunks invalid!",
+								"Unsupported color model other than RGB is specified in PNG!");
+					validChunkType = !(hasSeenPLTE || hasSeenIDAT);
+					hasSeenCICP = true;
+				}
+
+				if (!skip && "mDCV".equals(chunkTypeString)) {
+					validChunkType = !(hasSeenPLTE || hasSeenIDAT);
+					hasSeenMDCV = true;
+				}
+
+				if (!validChunkType) {
+					for (int i = 0; i < HARMLESS_CHUNK_TYPES_BEFORE_PLTE.length; i++) {
+						if (HARMLESS_CHUNK_TYPES_BEFORE_PLTE[i].equals(chunkTypeString)) {
+							if (!hasSeenPLTE && !hasSeenIDAT) {
+								validChunkType = true;
+								break;
+							} else {
+								throwError("The chunk appeared in an unexpected order!",
```

Comment:
One of those messages should mention _which_ chunk is out of order, so the user can potentially do something about it.